### PR TITLE
Fixes and improvements in the parsing of typed lists

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

This PR brings several fixes and improvements in the parsing of typed lists (names/variables).

## Fixes

n/a

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works

## Further comments

n/a

### CODE REVIEW

File: pddl/parser/domain.py [hunk 1]
```diff
@@ -289,47 +292,24 @@ def constant(self, args):
...
+        try:
+            types_index = TypesIndex.parse_typed_list(args)
+            return types_index.get_typed_list_of_names()
+        except ValueError as e:
+            raise PDDLParsingError(
+                f"error while parsing tokens {list(map(str, args))}: {str(e)}"
+            ) from None
 
-    def typed_list_variable(self, args) -> Dict[str, Set[str]]:
+    def typed_list_variable(self, args) -> OrderedDictType[name, Set[name]]:
```
Comment (marcofavorito) [inline pddl/parser/domain.py:312]: `OrderedDict` is essential, predicates variables and action parameters require an ordered (possibly typed) list of variables.

The same cannot be said for `forall`/`exists` expressions, but having them ordered does not harm.

File: pddl/parser/symbols.py [hunk 2]
```diff
@@ -21,34 +21,36 @@
 class Symbols(Enum):
```
Comment (marcofavorito) [inline pddl/parser/symbols.py:21]: just sorted alphanumerically.

File: pddl/parser/types_index.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    def parse_typed_list(cls, tokens: List[Union[str, List[str]]]) -> "TypesIndex":
+        """
+        Parse typed list.
+
+        This method takes in input a list of tokens as returned by the domain or problem parsers,
+        and returns a TypesIndex object.
+
+        The input list of tokens must have the following format:
+        - if the list is not typed, it is simply a list of names
+        - if the list is typed, the format is: [name_1, ..., name_n], "-", [type_1, ..., type_m], ...
```
Comment (francescofuggitti) [inline pddl/parser/types_index.py]: Shouldn't this be `[name_1, ..., name_n], "-", parent_type, ...` as below?

**Review by francescofuggitti — APPROVED** [review]
Looks good! Thank you 💯

File: pddl/parser/types_index.py (hunk 3 above)
Comment (marcofavorito) [reply pddl/parser/types_index.py]: outdated comment, please see the code

File: .github/workflows/docs.yml [hunk 4]
```diff
@@ -13,7 +13,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest]
-        python-version: [3.7]
+        python-version: ["3.8"]
 
     timeout-minutes: 30
```
Comment (francescofuggitti) [inline .github/workflows/docs.yml:19]: Shall we change:
```
- uses: actions/checkout@master
- uses: actions/setup-python@master
```

to `main` as well?

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -13,7 +13,7 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest]
-        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11"]
+        python-version: ["3.8", "3.9", "3.10", "3.11"]
 
     timeout-minutes: 30
```
Comment (francescofuggitti) [inline .github/workflows/test.yml:19]: same here?

File: .github/workflows/docs.yml (hunk 4 above)
Comment (marcofavorito) [reply .github/workflows/docs.yml:19]: Yes sure, thank you for spotting it

**Comment by ssardina** [comment]
hi @francescofuggitti @marcofavorito , I was about to submit a PR to fix the constant section the domain, as they currently do not come with their types.

However, I just found this massive PR that seems to deal with related things and will change many things, so don't want to cause conflicts and so I am planning to wait until this is merge. Is it almost done?

I think the problem is in `pddl.formatter`:

```
    body += _sort_and_print_collection("(:constants ", domain.constants, ")\n")
```

That can be improved I think to do the same as objects in problems... I think.

(BTW, great work on this parser, it's very clean!)

**Comment by marcofavorito** [comment]
> hi @francescofuggitti @marcofavorito , I was about to submit a PR to fix the constant section the domain, as they currently do not come with their types.
> 
> However, I just found this massive PR that seems to deal with related things and will change many things, so don't want to cause conflicts and so I am planning to wait until this is merge. Is it almost done?
> 
> I think the problem is in `pddl.formatter`:
> 
> ```
>     body += _sort_and_print_collection("(:constants ", domain.constants, ")\n")
> ```
> 
> That can be improved I think to do the same as objects in problems... I think.
> 
> (BTW, great work on this parser, it's very clean!)

Hi @ssardina , thank you for your appreciation... being useful to the community is the lymph for our work and motivation!

Given your message, we will finish the PR very very soon (~hours). We will keep you posted here :)

File: tests/test_formatter.py [hunk 6]
```diff
@@ -36,3 +40,62 @@ def test_problem_formatter(problem_parser, pddl_file):
...
+        constants=[a1, b1, c1, d2, e2, f2, g3, h3, i3, j, k, lc],
+    )
+
+    domain_str = domain_to_string(domain)
+
+    assert domain_str == dedent(
+        """\
+    (define (domain my_domain)
+        (:requirements :typing)
+        (:types type_2 type_3 - type_1 type_1)
```
Comment (marcofavorito) [inline tests/test_formatter.py:68]: Note, earlier we did not consider types with no parents, and we printed `type_1` in the first position. However, this is not correct wrt PDDL syntax; typed names with no type should be at the end of the list, o/w there is no way the parsing rules can spot them:
image

File: tests/test_formatter.py [hunk 7]
```diff
@@ -36,3 +40,62 @@ def test_problem_formatter(problem_parser, pddl_file):
...
+        (:constants a b c - type_1 d e f - type_2 g h i - type_3 j k l)
```
Comment (marcofavorito) [inline tests/test_formatter.py:69]: the same applies here: first group by type, print groups by type name in alphanumerical order, and then print constants sorted within the group.

File: tests/test_formatter.py [hunk 8]
```diff
@@ -36,3 +40,62 @@ def test_problem_formatter(problem_parser, pddl_file):
...
+        objects=[a1, b1, c1, d2, e2, f2, g3, h3, i3, j, k, lc],
+    )
+    problem_str = problem_to_string(problem)
+
+    assert problem_str == dedent(
+        """\
+    (define (problem problem-1)
+        (:domain my_domain)
+        (:requirements :typing)
+        (:objects a b c - type_1 d e f - type_2 g h i - type_3 j k l)
```
Comment (marcofavorito) [inline tests/test_formatter.py:97]: the same approach used for constants applies here.

File: tests/test_formatter.py [hunk 9]
```diff
@@ -36,3 +40,62 @@ def test_problem_formatter(problem_parser, pddl_file):
...
+        (:init )
+        (:goal (and))
```
Comment (marcofavorito) [inline tests/test_formatter.py:99]: we should decide what to do with empty init and goal sections. The PDDL grammar specifies these sections are mandatory. We will tackle this in a next PR.

**Comment by marcofavorito** [comment]
@ssardina this commit [edcc295](https://github.com/AI-Planning/pddl/pull/82/commits/edcc29547276e239fd945c7c827ac060b6636738) should solve your request. The formatting issue was planned for another PR, but it is even better that we've added it here.

@francescofuggitti, @haz, any concerns? Would you like to give a final review? I have other changes planned (integrate consistency checks, started [here](https://github.com/AI-Planning/pddl/tree/consistency-checks)), but I think this PR is already too big; so I will continue on another branch/PR.

**Comment by marcofavorito** [comment]
> 

Thank you for your reply @ssardina . In this issue #70, there has been a discussion regarding whether we should support them or not. The discussion converged to NOT supporting types with multiple parent types (I think it is summarized by this comment, that cites the PDDL textbook: https://github.com/AI-Planning/pddl/issues/70#issuecomment-1564858353)

I would be very glad to know your opinion on the topic though!

**Comment by marcofavorito** [comment]
Regarding the problem on `area`, thank you, it will be fixed very soon!

**Comment by marcofavorito** [comment]
However, in this PR I get an even more primitive error complaining that area is duplicated:
> 
> image
> 
> I believe a parent type can indeed be a child type in PDDL and is used in many domains, right? Like here `area`:
> 
> image

I think the problem in this case is that `area` occurs twice as child type, this is not allowed:

image


The error message will be changed so to be more informative.

**Review by francescofuggitti — APPROVED** [review]
Looks awesome! Thank you :)

**Comment by haz** [comment]
pedro-monkey-puppet

**Pull request merged**